u_roller.go

```diff
@@ -0,0 +1,110 @@
+package tls
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"net"
+	"sync"
+	"time"
+)
+
+type Roller struct {
+	HelloIDs            []ClientHelloID
+	HelloIDMu           sync.Mutex
+	NextHelloID         ClientHelloID
```

Comment:
I think this would be easier to understand if instead of `NextHelloID` you had `WorkingHelloID`, i.e. a known-working HelloID that is initially `nil`.  Then there are two clear states: either you have a known working ID (use it) or you don't (go down the list).  Transitions between states are also pretty clear.

Reply:
Done